src/Api/Upload.php

```diff
@@ -0,0 +1,170 @@
+
+			// validate chunk request
+			static::validateChunk($tmpWithId, $filename, $offset, $id);
+
+			// stream chunk content and append it to partial file
+			stream_copy_to_stream(
+				fopen($source, 'r'),
+				fopen($tmpWithId, 'a')
+			);
+
+			// clear file stat cache so the following call to `F::size`
+			// really returns the updated file size
+			clearstatcache();
+
+			// if partial file isn't complete yet, return early
+			if (F::size($tmpWithId) < (int)$total) {
+				return null;
+			}
+
+			// remove id from partial filename now the file is complete,
+			// so we can pass the path from the tmp upload directory
+			// as new source path for the file back to the API upload method
+			rename($tmpWithId, $tmp);
+
+			return $tmp;
+		}
+
+		return $source;
+	}
+
+	/**
+	 * Returns the ideal size for a file chunk
+	 */
+	public static function chunkSize(): int
+	{
+		$max = [
+			Str::toBytes(ini_get('upload_max_filesize')),
+			Str::toBytes(ini_get('post_max_size'))
+		];
+
+		// if server is behind a cloudflare proxy
+		if (isset($_SERVER['HTTP_CF_CONNECTING_IP']) === true) {
+			$max[] = Str::toBytes('100M');
+		}
+
+		// to be sure, only use 95% of the max possible upload size
+		return (int)floor(min($max) * 0.95);
+	}
+
+	/**
+	 * Clean up tmp directory of stale files
+	 */
+	public static function clean(): void
+	{
+		foreach (Dir::files($dir = static::tmp(), [], true) as $file) {
+			// remove any file that hasn't been altered
+			// in the last 24 hours
+			if (F::modified($file) < time() - 86400) {
+				F::remove($file);
+			}
+		}
+
+		if (Dir::isEmpty($dir) === true) {
+			Dir::remove($dir);
+		}
+	}
+
+	/**
+	 * Returns root of directory used for
+	 * temporarily storing (incomplete) uploads
+	 * @codeCoverageIgnore
+	 */
+	protected static function tmp(): string
+	{
+		return App::instance()->root('cache') . '/.uploads';
+	}
+
+	/**
+	 * Ensures the sent chunk is valid
+	 *
+	 * @throws \Kirby\Exception\DuplicateException Duplicate first chunk (same filename and id)
+	 * @throws \Kirby\Exception\Exception Chunk offset does not match existing tmp file
+	 * @throws \Kirby\Exception\NotFoundException Subsequent chunk has no  existing tmp file
+	 */
+	protected static function validateChunk(
+		string $tmp,
+		string $filename,
+		int $offset,
+		string $id
+	): void {
+		if (F::exists($tmp) === true) {
+			if ($offset === 0) {
+				throw new DuplicateException('A tmp file upload with the same filename "' . $filename . '" and upload id "' . $id . '" already exists');
+			}
+
+			if ($offset !== F::size($tmp)) {
+				throw new Exception('Chunk offset ' . $offset . ' does not match the existing tmp upload file size of ' . F::size($tmp));
+			}
+		} elseif ($offset > 0) {
+			throw new NotFoundException('Chunk offset ' . $offset . ' for non-existing tmp file "' . $filename . '"');
+		}
+
+		// validate file
+		$parent = new Page(['slug' => 'tmp']);
+		$file   = new File(['filename' => $filename, 'parent' => $parent]);
```

Comment:
This should operate on the actual filename as it's primarily a security check.

```suggestion
		$file   = new File(['filename' => basename($tmp), 'parent' => $parent]);
```

Reply:
@lukasbestle do you have an idea why this change causes the following failures?

```
There was 1 failure:

1) Kirby\Api\UploadTest::testValidateChunkInvalidExtension
Failed asserting that exception of type "TypeError" matches expected exception "Kirby\Exception\InvalidArgumentException". Message was: "stream_copy_to_stream(): Argument #1 ($from) must be of type resource, bool given" at
/home/runner/work/kirby/kirby/src/Api/Upload.php:70
/home/runner/work/kirby/kirby/tests/Api/UploadTest.php:346
.

--

1 test triggered 1 PHP warning:

Error: 1) /home/runner/work/kirby/kirby/src/Api/Upload.php:69
fopen(/home/runner/work/kirby/kirby/tests/tmp/Api.Upload/a.php): Failed to open stream: No such file or directory

Triggered by:

* Kirby\Api\UploadTest::testValidateChunkInvalidExtension
  /home/runner/work/kirby/kirby/tests/Api/UploadTest.php:331
  ```

Reply:
Does this change alone cause those failures? That's really weird as we are only doing a validation in this method and don't overwrite the value of `$tmp` that the calling method works with afterwards.

Reply:
Yes, it did for me. Just this change. I can test it again later this week if you say it doesn't happen for you when just changing this.

Reply:
Can't test it right now, but I can debug this further.

Reply:
Worked on the maxsize and shuffled things around. Now it uses `basename($tmp)` and works.